frontend/dockerfile/dockerfile2llb/convert.go

```diff
@@ -166,8 +166,25 @@ func Dockerfile2LLB(ctx context.Context, dt []byte, opt ConvertOpt) (*llb.State,
 			prefixPlatform: opt.PrefixPlatform,
 		}
 
+		if v := st.Platform; v != "" {
+			v, err := shlex.ProcessWordWithMap(v, metaArgsToMap(optMetaArgs))
+			if err != nil {
+				return nil, nil, nil, parser.WithLocation(errors.Wrapf(err, "failed to process arguments for platform %s", v), st.Location)
+			}
+
+			p, err := platforms.Parse(v)
+			if err != nil {
+				return nil, nil, nil, parser.WithLocation(errors.Wrapf(err, "failed to parse platform %s", v), st.Location)
+			}
+			ds.platform = &p
+		}
+
 		if st.Name != "" {
-			s, img, bi, err := opt.ContextByName(ctx, st.Name, opt.ImageResolveMode.String())
+			p := ds.platform
+			if p == nil {
+				p = opt.TargetPlatform
```

Comment:
I fixed the test, and the test fails without the patch, but I also expect it to fail if I always set `p` to `opt.TargetPlatform`, unfortunately it passes in that case which makes me think that maybe this option isn't what I think it is?

Reply:
>  if I always set p to opt.TargetPlatform

You mean even if `p != nil` ?

I wonder if it is somehow related to there being 2 modes for matching context. You can set `context:foo` but also `context:foo::linux/arm64` . Latter is for the case where you want to set a separate context for arch. If there is no arch in context name then any platform matches, but in your case it is not about matching context but pulling the right subimage from the matched context.

Reply:
Could you explain more about the case that was confusing to you. Maybe push a link to commit that has the invalid case where a test that should fail still passes.

Reply:
Yes, added a commit to show what I mean.
I'm expecting `opt.TargetPlatform` to be whatever was set via `--platform` (and thus basically ignoring all changes I'm making in this PR).

Reply:
Figured this out... the test isn't using a name for the stage so that part of the code was not even running.

Reply:
Or at least was doing this nil check in a place where it wasn't making a difference.